fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/plans/FDBComparatorPlanTest.java

```diff
@@ -0,0 +1,314 @@
+import org.junit.jupiter.api.Tag;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.apple.foundationdb.record.TestHelpers.assertDiscardedAtMost;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@Tag(Tags.RequiresFDB)
+public class FDBComparatorPlanTest extends FDBRecordStoreQueryTestBase {
+
+    @BeforeEach
+    void setup() throws Exception {
+    }
+
+    @DualPlannerTest
+    void testOneInnerPlan() throws Exception {
+        complexQuerySetup(NO_HOOK);
+        RecordQuery query = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query), primaryKey(), 0);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(1)),
+                context -> assertDiscardedAtMost(67, context));
+        assertEquals(33, count);
+    }
+
+    @DualPlannerTest
+    void testTwoSameInnerPlans() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        RecordQuery query1 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+        RecordQuery query2 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 0);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
+                record -> assertThat(record.getNumValue2(), is(1)),
+                context -> assertDiscardedAtMost(134, context));
+        assertEquals(33, count);
+    }
+
+    @DualPlannerTest
+    void testTwoDifferentInnerPlansIndex0() throws Exception {
+        complexQuerySetup(NO_HOOK);
+
+        RecordQuery query1 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(1))
+                .build();
+        RecordQuery query2 = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.field("num_value_2").equalsValue(2))
+                .build();
+
+        RecordQueryPlan planUnderTest = RecordQueryComparatorPlan.from(plan(query1, query2), primaryKey(), 0);
+
+        int count = querySimpleRecordStore(NO_HOOK, planUnderTest, EvaluationContext::empty,
```

Comment:
I take it that there should be logs of a discrepancy between the two plans, though it's a little hard to see that without either an explicit assertion on the logs (which would require being able to hook into the log system) or some other way of inspecting the cursor for errors detected. I guess something to think about

Reply:
There are ways to do this (e.g. https://stackoverflow.com/questions/3717402/how-can-i-test-with-junit-that-a-warning-was-logged-with-log4j), and there are also OSS that allows you to assert the contents of logs.
I would leave that for some later phase since we don't really do that anywhere else.